### ISSUE TITLE
clSetProgramSpecializationConstant description needs clarification

### ISSUE THREAD

**Issue by AlexeySachkov** [KhronosGroup/OpenCL-Docs#553]
At the moment, [`clSetProgramSpecializationConstant`](https://github.com/KhronosGroup/OpenCL-Docs/blob/master/api/opencl_runtime_layer.asciidoc#setting-spir-v-specialization-constants) only documents what happens if user uses `clCreateProgramWithIL` -> `clBuildProgram` sequence of APIs, however, it is possible to do separate compilation and linking even for SPIR-V inputs via `clCreateProgramWithIL` -> `clCompileProgram` -> `clBuildProgram`.

For the second usage model, the spec doesn't document which exact API call (`clCompileProgram` or `clBuildProgram`) uses values set by `clSetProgramSpecializationConstant`:

> The data pointed to by spec_value are copied and can be safely reused by the application after {clSetProgramSpecializationConstant} returns. This specialization value will be used by subsequent calls to {clBuildProgram} until another call to {clSetProgramSpecializationConstant} changes it. 

> Calling this function multiple times for the same specialization constant shall cause the last provided value to override any previously specified value. The values are used by a subsequent {clBuildProgram} call for the program.

**Comment by jenatali** [comment]
+1 to this. I'm looking at supporting SPIR-V + specialization constants for OpenCLOn12, and the outcome of this question would determine how I'd implement that.

If specialization constants are required to be consumed by `clCompileProgram`, that means I need to use SPIRV-Tools to do the specialization, as my compiled representation *is* SPIR-V. On the other hand, if specialization constants can only be used for already-linked SPIR-V that's used with `clBuildProgram`, my SPIR-V -> DXIL pipeline does have a way to specialize during translation, but only on fully-linked programs.

**Comment by AlexeySachkov** [comment]
> If specialization constants are required to be consumed by clCompileProgram, that means I need to use SPIRV-Tools to do the specialization, as my compiled representation is SPIR-V

I think this is the intended way

> On the other hand, if specialization constants can only be used for already-linked SPIR-V that's used with clBuildProgram, my SPIR-V -> DXIL pipeline does have a way to specialize during translation, but only on fully-linked programs.

Because in this case we would have to document how specialization constants are re-enumerated during linking, because it is possible that two incoming SPIR-V use overlapping set of IDs assigned to specialization constants and in order to set a spec constant value, user has to pass the ID to the corresponding API.

**Comment by jenatali** [comment]
Yeah, after thinking about it some more I agree with you and implemented that.

**Comment by bashbaug** [comment]
Blast from the past, I know, but I'd like to finally get this issue resolved.

I made some tests to drive discussion, see: https://github.com/KhronosGroup/OpenCL-CTS/pull/2774

There are a few things we should confirm:

1. Do we agree that the specialization constant value should be taken when the program is compiled vs. when the program is linked?  This seems to be the most common behavior and preferred resolution back in 2021.  This is the current behavior tested by the draft tests.
2. If a program is already compiled or linked or built, or said another way when the program binary type is not `BINARY_TYPE_NONE`, should it be an error to set the value of a specialization constant?  Although this seems like desired behavior, it may not be possible, since the spec strongly implies that it should be possible to `clSetProgramSpecializationConstant` -> `clBuildProgram` -> `clSetProgramSpecializationConstant` with a new value -> `clBuildProgram` -> etc.  The current draft tests expect that it is valid to call `clSetProgramSpecializationConstant` after compiling the program, but the new value is ignored (unless the program is compiled or built again - this last part does not appear to be covered by existing tests or the draft tests).
